python/smqtk/bin/train_itq.py

```diff
@@ -63,12 +84,25 @@ def uuids_iter():
             with open(uuids_list_filepath) as f:
                 for l in f:
                     yield l.strip()
+        uuids = uuids_iter()
         log.info("Loading UUIDs list from file: %s", uuids_list_filepath)
-        d_iter = descriptor_index.get_many_descriptors(uuids_iter())
+        if max_descriptors:
+            uuids = list(uuids)
+            if max_descriptors < len(uuids):
+                log.info("Subsampling UUIDs (old count=%d, new count=%d)",
+                         len(uuids), max_descriptors)
+                uuids = random.sample(uuids, max_descriptors)
+        d_iter = descriptor_index.get_many_descriptors(uuids)
     else:
+        d_length = len(descriptor_index)
         log.info("Using UUIDs from loaded DescriptorIndex (count=%d)",
-                 len(descriptor_index))
-        d_iter = descriptor_index
+                 d_length)
+        if max_descriptors and max_descriptors < d_length:
+            log.info("Subsampling loaded DescriptorIndex (new count=%d)",
+                     max_descriptors)
+            d_iter = subsample(descriptor_index, max_descriptors, d_length)
```

Comment:
Using ``numpy.random.choice`` here instead of implementing a new subsample function might be better.

Reply:
Wouldn't that require loading all the descriptors into memory? My intention was to avoid that, hence the streaming version. What I could do instead is `uuids = random.sample(list(descriptor_index.iterkeys()), max_descriptors)` and then use `get_many_descriptors`. That would presumably still have lower memory requirements, but I don't know off the top of my head which would be faster.

Reply:
You can use ``numpy.random.replace(d_length, max_descriptors, replace=False)`` to get an array of randomly sampled indices to select, then perform an iteration over descriptor index keys to select the appropriate keys, then using ``descriptor_index.get_many_descriptors`` to get the larger descriptor values for only those specific keys.

If you wanted to me more deterministic, seed-able approach but use more RAM, you could possibly load all keys into RAM, which should be much smaller than the actual descriptors, deterministically sort them and *then* do the random selection.

Reply:
Thanks for the suggestions. I found a succinct expression of the indexing idea and have added a commit to use that instead.